### PR TITLE
Eliminate the search_without_count function

### LINKED ISSUE #461: Do we need the search_without_count() functions?
**I'm submitting a ...**
- [ ] bug report
- [X] feature request

**Current behavior:**

We have many repetitive functions that I believe could be eliminated.

**Expected behavior:**

Fewer functions required to deliver the equivalent functionality.

**Steps to reproduce:**
Our scrapers have:
* [__def search(self, query, num_results, qtype=''):__](https://github.com/fossasia/query-server/blob/master/app/scrapers/generalized.py#L52) and
* [__def search_without_count(self, query):__](https://github.com/fossasia/query-server/blob/master/app/scrapers/generalized.py#L76)
but perhaps we could remove the second and only have the first.

When we do a __search()__ we are not guaranteeing that we will deliver at least __num_results__ results.  If the upstream search engine only delivers 5 results then we will only return that many even if the caller has set __num_results__ to 10.  If the upstream search engine does not allow us to control how many results we can get then we are not breaking any guarantee if we deliver less than __num_results__.  We are however guaranteeing that we will deliver no more than __num_results__.

So, why not have a default value for the num_results parameter (__def search(self, query, num_results=10, qtype=''):__) and do away with the _without_count_ functions?  Different scrapers could set their default value higher or lower.  The final return could ensure that no more than the requested number of urls are returned with `return urls[:num_results]`

Thoughts?

### PR DESCRIPTION
**Pull request by ankan17**
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:

The functionality of the search_without_count method has been
implemented within the search method and 3 tests have been added for
each scraper with a common search query "fossasia". The tests are:

 - Without any count
 - With a small count
 - With a large count

Resolves #461

### CODE REVIEW

**Comment by ankan17** [comment]
@realslimshanky I don't see any conflicts

**Comment by ankan17** [comment]
@realslimshanky @cclauss Please review the PR and suggest changes if required

File: test/test_ask.py [hunk 1]
```diff
@@ -74,3 +74,24 @@ def test_parse_video_response():
         }
     ]
     assert resp == expected_resp
+
+
+def test_search_ask_without_count():
+    query = 'fossasia'
+    expected_max_resp_count = 10
+    resp_count = len(Ask().search(query))
+    assert resp_count <= expected_max_resp_count
```
Comment (cclauss) [inline test/test_ask.py]: With a popular search term like _fossasia_ we should get all of the search results that we ask for even if we ask for 27.  My worry with the current test is that the scraper could return zero hits and still pass the test.

Comment (ankan17) [reply test/test_ask.py]: Quora returns only 9 results for "fossasia" and in general, the maximum response count for quora is 10. I don't know about the results of other search engines. So, should I reduce the max_count(27) and check for "==" and have different tests for quora?

Comment (cclauss) [reply test/test_ask.py]: Keep the high maxcount but keep the <= where it _needs_ to be but explain that choice in a comment.

Comment (ankan17) [reply test/test_ask.py]: Sorry for the late reply. I will make the changes by tomorrow.

Comment (ankan17) [reply test/test_ask.py]: @cclauss I have added the necessary changes, but can you help me understand why the travis ci build is failing for python2?

**Comment by cclauss** [comment]
@rupav Both @ankan17 and I have reviewed the logs and can not spot the issue.  Can you?

**Comment by ankan17** [comment]
@cclauss That is what I am saying, I saw the logs and couldn't understand what's the problem. I ran the tests in a python2 environment locally and the tests ran just fine.

**Comment by ankan17** [comment]
Due to lack of response, closing this PR.

**Comment by cclauss** [comment]
Can we please drop support for Python 2 as requested in #391 ?  There are a lot of frustrations with maintaining compatibility with both legacy Python and Python.  Unicode is the default in Python 3 so issues like #386 and this one plus many of the test cases and workarounds in the code would become much easier to deal with.  Let's join the crowd... http://www.python3statement.org

This is a server-based process and there are not many instance of it running in the world which means that dropping support for legacy Python should not effect many deployments.

**Pull request closed without merging**